app/models/cinstance.rb

```diff
@@ -111,17 +111,20 @@ def validate_plan_is_unique?
   validate :same_service, on: :update, if: :plan_id_changed?
 
   # letter, number, underscore (_), hyphen-minus (-), dot (.), base64 format
-  # In base64 encoding, the character set is [A-Z,a-z,0-9,and + /], if rest length is less than 4, fill of '=' character.
-  # ^([A-Za-z0-9+/]{4})* means the String start with 0 time or more base64 group.
-  # ([A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==) means the String must end of 3 forms in [A-Za-z0-9+/]{4} or [A-Za-z0-9+/]{3}= or [A-Za-z0-9+/]{2}==
+  # In base64 encoding, the character set is [A-Z,a-z,0-9,and +], if rest length is less than 4, fill of '=' character.
+  # ^([A-Za-z0-9+]{4})* means the String start with 0 time or more base64 group.
+  # ([A-Za-z0-9+]{4}|[A-Za-z0-9+]{3}=|[A-Za-z0-9+]{2}==) means the String must end of 3 forms in [A-Za-z0-9+]{4} or [A-Za-z0-9+]{3}= or [A-Za-z0-9+]{2}==
   # matches also the non 64B case with (\A[\w\-\.]+\Z)
-  USER_KEY_FORMAT = /(([\w\-\.]+)|([A-Za-z0-9+\/]{4})*([A-Za-z0-9+\/]{4}|[A-Za-z0-9+\/]{3}=|[A-Za-z0-9+\/]{2}==))/
-
-  validates :application_id, format: { with: /\A[\x20-\x7E]+\Z/ }, length: { in: 4..255 }
+  # NOTE: base64 format also accepts forward slash (/), however we don't allow it because of the restriction on the backend
+  USER_KEY_FORMAT = /(([\w\-.]+)|([A-Za-z0-9+]{4})*([A-Za-z0-9+]{4}|[A-Za-z0-9+]{3}=|[A-Za-z0-9+]{2}==))/.freeze
 
+  # The following characters are accepted:
+  # A-Z a-z 0-9 ! " # $ % & ' ( ) * + , - . : ; < = > ? @ [ \ ] ^ _ ` { | } ~
+  # Spaces and / are not allowed
+  validates :application_id, format: { with: /\A[\x21-\x2E\x30-\x7E]+\Z/ }, length: { in: 4..255 }
 
   validates :user_key, format: { with: /\A#{USER_KEY_FORMAT}\Z/ }, length: { maximum: 256 },
-            allow_nil: true, allow_blank: true
```

Comment:
How could a user key be `nil` before?

Reply:
It still can be, because `allow_blank` includes `nil` as well...

I will check if there are such entries in the DB. In theory those applications that are on App Id / App Key or OpenID modes don't need `user_key`, but I think technically they would still be auto-generated.

Reply:
Actually, I'm thinking... Maybe we need to stop allowing blank values, I think it doesn't make any sense. And actually, there are not so many blank user keys in the current DB, about 9 on the accounts that are still active, and among these there are at least 3 that belong to internal test accounts...

So, I think we can remove `allow_blank` as well.